docs/src/index.md

```diff
@@ -4,6 +4,350 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+
+### Self derivative `Δself`, `∂self` etc.
+
+!!! terminology
+    To my knowledge there is no standard terminology for this.
+    Other good names might be `Δinternal`/`∂internal`.
+
+From the mathematical perspective, one may have been wondering what all this `Δself`, `∂self` is.
+After all a function with two inputs, say `f(a, b)`, only has two partial derivatives:
+``\dfrac{∂f}{∂a}``, ``\dfrac{∂f}{∂b}``.
+Why then does a `pushforward` take in this extra `Δself`, and why does a `pullback` return this extra `∂self`?
+
+The reason is that in Julia the function `f` may itself have internal fields.
+For example a closure has the fields it closes over; a callable object (i.e. a functor) like a `Flux.Dense` has the fields of that object.
+
+**Thus every function is treated as having the extra implicit argument `self`, which captures those fields.**
+So every `pushforward` takes in an extra argument, which is ignored unless the original function had fields.
+In is common to write `function foo_pushforward(_, Δargs...)` in the case when `foo` does not have fields.
+Similarly every `pullback` return an extra `∂self`, which for things without fields is the constant `NO_FIELDS`, indicating there are no fields within the function itself.
+
+#### Pushforward / Pullback summary
+
+- **Pushforward:**
+    - returned by `ffrule`
+    - takes input space wiggles, gives output space wobbles
+    - 1 argument per original function argument + 1 for the function itself
+    - 1 return per original function return
+
+- **Pullback**
+   - returned by `rrule`
+   - takes output space wobbles, gives input space wiggles
+   - 1 argument per original function return
+   - 1 return per original function argument + 1 for the function itself
+
+#### Pushforward/Pullback and Total Derivative/Gradient
+
+The most trivial use of `frule` and returned `pushforward` is to calculate the [Total Derivative](https://en.wikipedia.org/wiki/Total_derivative):
+
+```julia
+y, f_pushforward = frule(f, a, b, c)
+ẏ = f_pushforward(1, 1, 1, 1)  # for appropriate `1`-like perturbation.
+```
+
+Then we have that `ẏ` is the _total derivative_ of `f` at `(a, b, c)`, written mathematically as ``df_{(a,b,c)}``
+
+Similarly, the most trivial use of `rrule` and returned `pullback` is to calculate the [Gradient](https://en.wikipedia.org/wiki/Gradient):
+
+```julia
+y, f_pullback = rrule(f, a, b, c)
+∇f = f_pullback(1)  # for appropriate `1`-like seed.
+s̄, ā, b̄, c̄ = ∇f
+```
+Then we have that `∇f` is the _gradient_ of `f` at `(a, b, c)`.
+And we thus have the partial derivatives ``f̄ = \dfrac{∂f}{∂f}``, ``ā` = \dfrac{∂f}{∂a}``, ``b̄ = \dfrac{∂f}{∂b}``, ``c̄ = \dfrac{∂f}{∂c}``, including the and the self-partial derivative, ``f̄``.
+
+### Differentials
+
+The values that come back from pullbacks or pushforwards are not always the same type as the input/outputs of the original function.
+They are differentials; differency-equivalents.
```

Comment:
Are you sure "differency" is the word you want to use here? It doesn't seem to be used frequently in this context https://www.collinsdictionary.com/dictionary/english/differency .

Reply:
It is a word I made up. It definately is not a word that should be here.
I just needed to get something down.

I want to capture the idea that this is the result of some difference expression.
That is to say it doesn't need to be the same type as the input,
I guess the throw math at it,
For some domain 𝔽,
the difference-ish domain 𝔻 is a domain that supports
```
- : 𝔽×𝔻 → 𝔽
+ : 𝔻×𝔻 → 𝔻
* : ℝ×𝔻 → 𝔻
```

Interestly, that isn't exactly what we implement, but at the conceptual level that is what it is.

Reply:
I think it should be 
    - : 𝔽×𝔽 → 𝔻
right? And probably also
    + : 𝔽×𝔻 → 𝔽

This looks like a rough description of the fact that total derivative at a point is a linear function: https://en.wikipedia.org/wiki/Total_derivative . Linear functions can be added, subtracted and multiplied by a scalar.

Reply:
`- : 𝔽×𝔽 → 𝔻` is basically how we define `𝔻`, but I don't know that we need it.
But we do need `- : 𝔽×𝔻 → 𝔽` to be able to do gradient desent.
I suspect combined with some other feature we need that I am neglecting to mention we get that.

I was thinking that.
`+ : 𝔽×𝔻 → 𝔽` emerges as a consiquence of `- : 𝔽×𝔻 → 𝔽` and `* : ℝ×𝔻 → 𝔻`
and `-1∈ℝ`
and some properties of what `*` has to do.

> This looks like a rough description of the fact that total derivative at a point is a linear function:

Seems right, yes.

Reply:
So anyway, what is a nice short-way and handway of capturing that differentials have these properties.

Reply:
> `+ : 𝔽×𝔻 → 𝔽` emerges as a consiquence of `- : 𝔽×𝔻 → 𝔽` and `* : ℝ×𝔻 → 𝔻`
> and `-1∈ℝ`

OK, I understand now what you mean.

> So anyway, what is a nice short-way and handway of capturing that differentials have these properties.

Maybe you could write that `+` and `*` make differentials at a certain point a linear space (or vector space). It is less vague than "which let them act as mathematical objects".

Reply:
Its actually not a vector space.
The opterators they actually have are weirder still.
(like it is pretty odd you can multiply them.)

Reply:
I think for this section at least I am happy with what is there now.
Which talks about them representing the difference of two things of the original type

We might want more later for a more detailed page on differentials

Reply:
Well, at least 𝔻 is a vector space, right? There is of course a lot more structure there.